supply/spec/uploader_spec.rb

```diff
@@ -221,5 +221,200 @@ def find_obbs
+            image_type = args[:image_id].split('_')[1]
+            final_remote_images_ids[image_type].delete(args[:image_id])
+          end
+          allow(client).to receive(:upload_image) do |**args|
+            path = File.basename(args[:image_path])
+            image_type = args[:image_type]
+            final_remote_images_ids[image_type] << "new-id_#{image_type}_#{path}"
+          end
+
+          Supply::SCREENSHOT_TYPES.each do |screenshot_type|
+            remote_images = local_images.map do |path|
+              Supply::ImageListing.new("id_#{screenshot_type}_#{path}", '_unused_', "sha256-of-#{path}", '_unused_')
+            end # remote images will be in order 0124356 though
+            allow(client).to receive(:fetch_images).with(image_type: screenshot_type, language: language).and_return(remote_images)
+
+            final_remote_images_ids[screenshot_type] = remote_images.map(&:id)
+
+            # We should skip image0, image1, image2 from remote as they are the same as the first local images,
+            # But also skip image3 (which was in-between 2 and 3 in remote listing, but is still present in local images)
+            # While deleting image4 (because it was in-between image2 and image3 in the `remote_images`, so out of order)
+            # And finally deleting image5 and image6, before re-uploading image4, image5 and image6 in the right order
+            local_images.sort[0..3].each do |path|
+              expect(client).not_to receive(:clear_screenshot).with(image_type: screenshot_type, language: language, image_id: "id_#{screenshot_type}_#{path}")
+              expect(client).not_to receive(:upload_image).with(image_path: File.expand_path(path), image_type: screenshot_type, language: language)
+            end
+            local_images.sort[4..6].each do |path|
+              expect(client).to receive(:clear_screenshot).with(image_type: screenshot_type, language: language, image_id: "id_#{screenshot_type}_#{path}")
+              expect(client).to receive(:upload_image).with(image_path: File.expand_path(path), image_type: screenshot_type, language: language)
+            end
+          end
+
+          uploader = Supply::Uploader.new
+          uploader.upload_screenshots(language)
+
+          # Check the final order of the remote images after the whole skip/delete/upload dance
+          Supply::SCREENSHOT_TYPES.each do |screenshot_type|
+            expected_final_images_ids = %W[
+              id_#{screenshot_type}_image0.png
+              id_#{screenshot_type}_image1.png
+              id_#{screenshot_type}_image2.png
+              id_#{screenshot_type}_image3.png
+              new-id_#{screenshot_type}_image4.png
+              new-id_#{screenshot_type}_image5.png
+              new-id_#{screenshot_type}_image6.png
+            ]
+            expect(final_remote_images_ids[screenshot_type]).to eq(expected_final_images_ids)
+          end
+        end
```

Comment:
This unit test is the most interesting case I wanted to get right, as it shows a case where you have the same images locally and remotely, but you swapped the order of some of your screenshots:

 - On the Play Store you initially had the images that correspond to `image0.png`, `image1.png`, `image2.png`, `image4.png`, `image3.png`, `image5.png`, `image6.png`, currently appearing in that order in the Play Store
 - But since then, you decided to swap the order of `image4.png` and `image3.png` and now you want to re-upload the images (which will be uploaded in alphabetical order, as `supply` has always done)
 - This will thus:
    - Keep `image0.png`, `image1.png` and `image2.png` untouched
    - Delete the `image4.png`, which is currently after `image2.png` in PlayStore — because it's now out of order compared to the local files
    - But keep `image3.png` in PlayStore (that came after `image4.png`, but will now come after `image2.png` since `image4.png` was deleted from the PlayStore edit)
    - And then delete `image5.png` and `image6.png`, to finally re-upload `image4.png`, `image5.png` and `image6.png` in the correct order.

This looks like this when running this spec with `DEBUG=1 bundle exec rspec supply/spec/uploader_spec.rb`:
```
…
[22:30:47]: 🔍 Checking phoneScreenshots checksums...
[22:30:47]: 🟰 Skipping upload of screenshot image0.png as remote sha256 matches.
[22:30:47]: 🟰 Skipping upload of screenshot image1.png as remote sha256 matches.
[22:30:47]: 🟰 Skipping upload of screenshot image2.png as remote sha256 matches.
[22:30:47]: 🚮 Deleting pt-BR screenshot id #id_phoneScreenshots_image4.png as it does not exist locally or is out of order...
[22:30:47]: 🟰 Skipping upload of screenshot image3.png as remote sha256 matches.
[22:30:47]: 🚮 Deleting pt-BR screenshot id #id_phoneScreenshots_image5.png as it does not exist locally or is out of order...
[22:30:47]: 🚮 Deleting pt-BR screenshot id #id_phoneScreenshots_image6.png as it does not exist locally or is out of order...
[22:30:47]: ⬆️  Uploading screenshot image4.png...
[22:30:47]: ⬆️  Uploading screenshot image5.png...
[22:30:47]: ⬆️  Uploading screenshot image6.png...
…
```